### PR TITLE
Fix memory leak in xds_proxy when channels are blocked

### PR DESCRIPTION
**Pull request by howardjohn**
Fixes https://github.com/istio/istio/issues/34097

~I have some tests locally but its really hacky, will cleanup and add before submitting~ Tests added. This was a bit tricky to test, but the test fails without the PR and passes with it, so I think its all good.

### CODE REVIEW

**Review by lei-tang — APPROVED** [review]

**Review by lei-tang — COMMENTED** [review]
Thanks for the PR! There are multiple test failures (unit test, lint test, etc). Please fix them.

File: pilot/test/xdstest/grpc.go [hunk 1]
```diff
@@ -0,0 +1,74 @@
...
+	}
+}
+
+type slowClientStream struct {
+	grpc.ClientStream
+	recv, send time.Duration
+}
+
+func (w *slowClientStream) RecvMsg(m interface{}) error {
+	if w.recv > 0 {
```
Comment (lei-tang) [inline pilot/test/xdstest/grpc.go:25]: Where is w.recv updated? Does it need a mutex protection?

Comment (howardjohn) [reply pilot/test/xdstest/grpc.go:25]: Its not mutated, just set once at construction

**Comment by howardjohn** [comment]
Thanks, removed my logs

File: pkg/istio-agent/xds_proxy.go [hunk 2]
```diff
@@ -282,6 +289,15 @@ type ProxyConnection struct {
 	upstreamDeltas     discovery.AggregatedDiscoveryService_DeltaAggregatedResourcesClient
 }
 
+// safeSendRequest is a small wrapper around sending to con.requestsChan. This ensures that we do not
+// block forever on
```
Comment (stevenctl) [inline pkg/istio-agent/xds_proxy.go:293]: nit: "block forever on a full requestsChan"?

**Review by stevenctl — APPROVED** [review]

File: pkg/istio-agent/xds_proxy.go [hunk 3]
```diff
@@ -282,6 +289,15 @@ type ProxyConnection struct {
...
+func (con *ProxyConnection) safeSendRequest(req *discovery.DiscoveryRequest) {
```
Comment (ramaraochavali) [inline pkg/istio-agent/xds_proxy.go:294]: nit: We can just probably call this `sendRequest`

**Comment by hzxuzhonghu** [comment]
/cherrypick release-1.10


It seems backport to 1.9 and 1.10 failed

**Comment by hzxuzhonghu** [comment]
sorry, it is done manually

**Pull request merged**